From 89bd00b2b1f3628b4a6f4c3630f0a1010a721149 Mon Sep 17 00:00:00 2001
From: Eric Ruei <e-ruei1@ti.com>
Date: Mon, 19 Mar 2018 11:12:41 -0400
Subject: [PATCH 199/199] net: ethernet: ti: iep: add PPS offset support for
 PRUICSS

This patch records the PPS offset value and then adjust the PPS output
time based on that.

Signed-off-by: Eric Ruei <e-ruei1@ti.com>
---
 drivers/net/ethernet/ti/iep.c | 6 ++++++
 drivers/net/ethernet/ti/iep.h | 1 +
 2 files changed, 7 insertions(+)

diff --git a/drivers/net/ethernet/ti/iep.c b/drivers/net/ethernet/ti/iep.c
index c0ce982..1576733 100644
--- a/drivers/net/ethernet/ti/iep.c
+++ b/drivers/net/ethernet/ti/iep.c
@@ -227,6 +227,7 @@ static int iep_adjfreq(struct ptp_clock_info *ptp, s32 ppb)
 	/* if at least one of the pps is enabled, update cmp accordingly. */
 	if ((iep->pps[0].enable == 1) || (iep->pps[1].enable == 1)) {
 		ns_to_sec = NSEC_PER_SEC - ts.tv_nsec;
+		ns_to_sec += iep->pps[0].offset;
 		cyc_to_sec = iep_ns2cyc(iep, ns_to_sec);
 
 		/* +++TODO: fine tune the randomly fixed 10 ticks */
@@ -380,6 +381,7 @@ static int iep_pps_enable(struct iep *iep, unsigned int pps, int on)
 
 	/* align cmp count to next sec boundary */
 	ns_to_sec_bd = NSEC_PER_SEC - ts.tv_nsec;
+	ns_to_sec_bd += iep->pps[pps].offset;
 	cyc_to_sec_bd = iep_ns2cyc(iep, ns_to_sec_bd);
 	cmp_val = iep->tc.cycle_last + cyc_to_sec_bd;
 
@@ -514,6 +516,9 @@ static int iep_ptp_feature_enable(struct ptp_clock_info *ptp,
 		}
 
 		return iep_pps_enable(iep, IEP_PPS_EXTERNAL, on);
+	case PTP_CLK_REQ_PPS_OFFSET:
+		iep->pps[IEP_PPS_INTERNAL].offset = on;
+		return 0;
 
 	default:
 		break;
@@ -682,6 +687,7 @@ static long iep_overflow_check(struct ptp_clock_info *ptp)
 	 * out in the next check.
 	 */
 	ns_to_sec = NSEC_PER_SEC - ts.tv_nsec;
+	ns_to_sec += iep->pps[0].offset;
 	cyc_to_sec = iep_ns2cyc(iep, ns_to_sec);
 	cmp_val = iep->tc.cycle_last + cyc_to_sec;
 
diff --git a/drivers/net/ethernet/ti/iep.h b/drivers/net/ethernet/ti/iep.h
index ce33500..4820e0d 100644
--- a/drivers/net/ethernet/ti/iep.h
+++ b/drivers/net/ethernet/ti/iep.h
@@ -98,6 +98,7 @@ struct pps {
 	struct pinctrl_state *pin_on;
 	struct pinctrl_state *pin_off;
 	int enable;
+	int offset;
 	int next_op;
 	enum {
 		OP_DISABLE_SYNC,
-- 
1.9.1

